### PR TITLE
automate-dex: support different name id policy formats

### LINKED ISSUE #937: expose SAML NameIDPolicyFormat config option
## User Story

As an Automate User, I will use this work when I want to configure Automate authentication with SAML.

Problem: 
From @andy-dufour in slack:
```
Hi all, having some issues getting saml setup with a customer, 
they're having some issues with the name field: 
"We are providing name identifier in the this format: 
urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified 

and Chef Automate is expecting in this format: 
urn:oasis:names:tc:SAML:2.0:nameid-format:persistent. 
We do not have the option to modify it."
```

## Definition of Done
- SAML can be optionally configured with a [NameIDPolicyFormat](https://github.com/dexidp/dex/blob/master/connector/saml/saml.go#L115) with [these input options](https://github.com/dexidp/dex/blob/master/connector/saml/saml.go#L27-L35)

## Demo Script / Repro Steps
1. save the following config in a `test.toml` file
```
[dex.v1.sys.connectors.saml]
name_id_policy_format = "unspecified" // or any other option dex accepts
```
2. run 
``` 
chef-automate config patch test.toml
```
3. If we can get more details about the customer's failing config, we could try to repro their issue and see if this config patch fixes it

### PR DESCRIPTION
**Pull request by srenatus**
### :nut_and_bolt: Description

With this change, the "name id policy format" A2's SAML integration expects can be configured. Before, it was basically hardcoded to `urn:oasis:names:tc:SAML:2.0:nameid-format:persistent`.

- [X] expose dex' config option
- [X] validate allowed settings
- [X] configure default when unset (previously-hardcoded value)
- [x] document this

### :athletic_shoe: Demo Script / Repro Steps

- `start_all_services`, `rebuild components/automate-deployment`,`rebuild components/automate-cli`, `rebuild components/automate-dex`
- check the dev env's SAML config:
```
[7][default:/src:0]# grep -A10 SAML /hab/svc/automate-dex/var/etc/config.yml
    name: SAML
    config:
      ssoURL: https://chef.okta.com/app/chefsoftware_a2localfreshinstallunstable_1/exk1d6ztz4rioEOYA1d8/sso/saml
      ca: /hab/svc/automate-dex/config/saml-ca.pem
      redirectURI: https://a2-dev.test/dex/callback
      entityIssuer: https://automate.chef.co/dex/callback
      usernameAttr: email
      emailAttr: email
      groupsAttr: groups
      nameIDPolicyFormat: urn:oasis:names:tc:SAML:2.0:nameid-format:persistent
```
Note that this is the default in action, it's not displayed in `chef-automate config show` until you set it to something:
```
[8][default:/src:0]# chef-automate config show | grep -A7 saml
        [dex.v1.sys.connectors.saml]
          ca_contents = "-----BEGIN CERTIFICATE-----...-----END CERTIFICATE-----"
          sso_url = "https://chef.okta.com/app/chefsoftware_a2localfreshinstallunstable_1/exk1d6ztz4rioEOYA1d8/sso/saml"
          email_attr = "email"
          username_attr = "email"
          groups_attr = "groups"
          entity_issuer = "https://automate.chef.co/dex/callback"
[9][default:/src:0]# cat nameid_policy.toml
        [dex.v1.sys.connectors.saml]
name_id_policy_format = "urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified"
[10][default:/src:0]# chef-automate config patch nameid_policy.toml

Updating deployment configuration

Applying deployment configuration
  Installed automate-cli
  Started automate-dex
Success: Configuration patched
[11][default:/src:0]# chef-automate config show | grep -A7 saml
        [dex.v1.sys.connectors.saml]
          ca_contents = "-----BEGIN CERTIFICATE-----....-----END CERTIFICATE-----"
          sso_url = "https://chef.okta.com/app/chefsoftware_a2localfreshinstallunstable_1/exk1d6ztz4rioEOYA1d8/sso/saml"
          email_attr = "email"
          username_attr = "email"
          groups_attr = "groups"
          entity_issuer = "https://automate.chef.co/dex/callback"
          name_id_policy_format = "urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified"

[12][default:/src:0]# grep nameID /hab/svc/automate-dex/var/etc/config.yml
      nameIDPolicyFormat: urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified
[13][default:/src:0]#
```

### :chains: Related Resources

Fixes #937.

### :white_check_mark: Checklist

- [X] Necessary tests added/updated?
- [x] Necessary docs added/updated?
- [X] Code actually executed?
- [X] Vetting performed (unit tests, lint, etc.)?

### CODE REVIEW

File: api/config/dex/config_request.go [hunk 1]
```diff
@@ -31,24 +31,6 @@ func NewConfigRequest() *ConfigRequest {
 	}
 }
 
-// NewLdapConnector returns a new instance of ConfigRequest_Ldap with zero
```
Comment (srenatus) [inline api/config/dex/config_request.go]: dead code

File: api/config/dex/config_request_test.go [hunk 2]
```diff
@@ -26,7 +26,7 @@ func completeMSADWithoutOverrides() *dex.ConfigRequest_V1_Msad_Ldap {
 func TestValidate(t *testing.T) {
 	t.Run("Validates when there are no connectors", func(t *testing.T) {
 		cfg := dex.DefaultConfigRequest()
-		assert.Nil(t, cfg.Validate())
+		assert.NoError(t, cfg.Validate())
```
Comment (srenatus) [inline api/config/dex/config_request_test.go]: drive-by cleanup now that `cfg.Validate()` returns something of interface `error`

File: api/config/dex/config_request.go (hunk 1 above)
Comment (blakestier) [reply api/config/dex/config_request.go]: 💥

File: api/config/dex/config_request.go [hunk 3]
```diff
@@ -253,3 +255,12 @@ func checkCertsPEM(cfgErr *shared.InvalidConfigError, key, data string) {
 		}
 	}
 }
+
+func (samlCfg *ConfigRequest_V1_Saml) setNameIDPolicyDefault() {
+	if samlCfg == nil {
+		return
+	}
```
Comment (msorens) [inline api/config/dex/config_request.go]: What makes this code unique, in that it requires checking for the zero-object of the type? I cannot recall ever seeing this check in all the auth methods we've implemented. (That is, clearly if we got a nil here we would not want to continue. But wouldn't that be true in lots of methods in auth code, too?)

Comment (srenatus) [reply api/config/dex/config_request.go]: Not unique, just a different use case: this is called as
```go
c.V1.Sys.GetConnectors().GetSaml().setNameIDPolicyDefault()
```
and the `nil` check allows us *not* to check if the saml connector is actually set -- but **if it is**, we'll add the default value.
